diff --git a/test/common/memory_usage.h b/test/common/memory_usage.h
index cf8b4180..af9bf35c 100644
--- a/test/common/memory_usage.h
+++ b/test/common/memory_usage.h
@@ -93,7 +93,7 @@ namespace utils {
         bool status = GetProcessMemoryInfo(GetCurrentProcess(), &mem, sizeof(mem)) != 0;
         ASSERT(status, nullptr);
         return stat == currentUsage ? mem.PagefileUsage : mem.PeakPagefileUsage;
-#elif __unix__
+#elif __unix__ && !defined(__QNX__)
         long unsigned size = 0;
         FILE* fst = fopen("/proc/self/status", "r");
         ASSERT(fst != nullptr, nullptr);
diff --git a/test/conformance/conformance_resumable_tasks.cpp b/test/conformance/conformance_resumable_tasks.cpp
index 70fd878d..79313eb5 100644
--- a/test/conformance/conformance_resumable_tasks.cpp
+++ b/test/conformance/conformance_resumable_tasks.cpp
@@ -16,7 +16,8 @@
 
 #include "common/test.h"
 
-#if (!__TBB_WIN8UI_SUPPORT && !defined(WINAPI_FAMILY) && !__ANDROID__)
+#if (!__TBB_WIN8UI_SUPPORT && !defined(WINAPI_FAMILY) && !__ANDROID__ \
+    && !__QNX__)
 
 #include "oneapi/tbb/task.h"
 #include "oneapi/tbb/task_group.h"
diff --git a/test/tbb/test_eh_thread.cpp b/test/tbb/test_eh_thread.cpp
index a8885e66..11c57068 100644
--- a/test/tbb/test_eh_thread.cpp
+++ b/test/tbb/test_eh_thread.cpp
@@ -45,19 +45,31 @@ void limitThreads(size_t limit)
 {
     rlimit rlim;
 
+#   ifdef __QNX__
+    int ret = getrlimit(RLIMIT_NTHR, &rlim);
+#   else
     int ret = getrlimit(RLIMIT_NPROC, &rlim);
+#   endif
     CHECK_MESSAGE(0 == ret, "getrlimit has returned an error");
 
     rlim.rlim_cur = (rlim.rlim_max == (rlim_t)RLIM_INFINITY) ? limit : utils::min((rlim_t)limit, rlim.rlim_max);
 
+#   ifdef __QNX__
+    ret = setrlimit(RLIMIT_NTHR, &rlim);
+#   else
     ret = setrlimit(RLIMIT_NPROC, &rlim);
+#   endif
     CHECK_MESSAGE(0 == ret, "setrlimit has returned an error");
 }
 
 size_t getThreadLimit() {
     rlimit rlim;
 
+#   ifdef __QNX__
+    int ret = getrlimit(RLIMIT_NTHR, &rlim);
+#   else
     int ret = getrlimit(RLIMIT_NPROC, &rlim);
+#   endif
     CHECK_MESSAGE(0 == ret, "getrlimit has returned an error");
     return rlim.rlim_cur;
 }
