diff --git a/test/cdr/ResizeTest.cpp b/test/cdr/ResizeTest.cpp
index 559e948..80666ab 100644
--- a/test/cdr/ResizeTest.cpp
+++ b/test/cdr/ResizeTest.cpp
@@ -25,6 +25,7 @@
 using namespace eprosima::fastcdr;
 using namespace ::exception;
 
+namespace eprosima::fastcdr::test {
 static const uint8_t octet_t = 32;
 static const char char_t =  'Z';
 static const wchar_t wchar =  'Z';
@@ -4906,3 +4907,4 @@ TEST(CDRResizeTests, ReserveBuffer)
     EXPECT_EQ(false, buffer2.reserve(100));
     EXPECT_EQ(10u, buffer2.getBufferSize());
 }
+}
\ No newline at end of file
diff --git a/test/cdr/SimpleTest.cpp b/test/cdr/SimpleTest.cpp
index 38e8b6a..31ddb0b 100644
--- a/test/cdr/SimpleTest.cpp
+++ b/test/cdr/SimpleTest.cpp
@@ -31,6 +31,7 @@ using namespace ::exception;
 #define BUFFER_LENGTH 2000
 #define N_ARR_ELEMENTS 5
 
+namespace eprosima::fastcdr::test {
 static const uint8_t octet_t = 32;
 static const char char_t =  'Z';
 static const wchar_t wchar = 'Z';
@@ -7137,3 +7138,4 @@ TEST(FastCDRTests, EmptyStringSerializationSize)
     });
     EXPECT_EQ(cdr_ser.get_serialized_data_length(), 5u);
 }
+}
