diff --git a/CMakeLists.txt b/CMakeLists.txt
index d5941243d66..aa1772834b5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -27,6 +27,14 @@ endif()
 # ---[ Project and semantic versioning.
 project(Torch CXX C)
 
+message(STATUS "CMAKE_CXX_COMPILER_ID is ${CMAKE_CXX_COMPILER_ID}")
+
+# Enable whole archive for CXX on QNXOS
+if(CMAKE_SYSTEM_NAME STREQUAL "QNX")
+set(CMAKE_CXX_LINK_LIBRARY_USING_WHOLE_ARCHIVE "LINKER:--whole-archive,<LINK_ITEM>,--no-whole-archive")
+set(CMAKE_CXX_LINK_LIBRARY_USING_WHOLE_ARCHIVE_SUPPORTED TRUE)
+endif()
+
 # When built via scikit-build-core, Python_EXECUTABLE is already set.
 # For standalone cmake invocations, find it via find_package.
 find_package(Python COMPONENTS Interpreter REQUIRED)
@@ -186,7 +194,7 @@ set(CPU_RISCV OFF)
 
 if(CMAKE_SYSTEM_PROCESSOR MATCHES "(AMD64|x86_64)")
   set(CPU_INTEL ON)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64|arm64)")
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64|arm64|aarch64le)")
   set(CPU_AARCH64 ON)
 elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(ppc64le)")
   set(CPU_POWER ON)
@@ -915,9 +923,9 @@ include(ExternalProject)
 
 # ---[ Dependencies ---[ FBGEMM doesn't work on x86 32bit and
 # CMAKE_SYSTEM_PROCESSOR thinks its 64bit
-if(USE_FBGEMM AND NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^(x86_64|aarch64)$")
+if(USE_FBGEMM AND NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^(x86_64|aarch64|aarch64le)$")
   message(WARNING
-    "x86_64 or aarch64 required for FBGEMM. "
+    "x86_64 or aarch64/le required for FBGEMM. "
     "Not compiling with FBGEMM. "
     "Turn this warning off by USE_FBGEMM=OFF.")
   set(USE_FBGEMM OFF)
@@ -1028,7 +1036,7 @@ endif()
 
 # Enable sleef on Arm(R) architecture by default (except Android)
 if((NOT ${CMAKE_SYSTEM_NAME} STREQUAL "Android")
-  AND("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "aarch64"))
+  AND("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "aarch64|aarch64le"))
   string(APPEND CMAKE_CXX_FLAGS " -DAT_BUILD_ARM_VEC256_WITH_SLEEF")
   add_definitions(-DAT_BUILD_ARM_VEC256_WITH_SLEEF)
 endif()
@@ -1183,7 +1191,7 @@ if(NOT MSVC)
   endif()
   append_cxx_flag_if_supported("-Wno-maybe-uninitialized" CMAKE_CXX_FLAGS)
   append_cxx_flag_if_supported("-fstandalone-debug" CMAKE_CXX_FLAGS_DEBUG)
-  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64" AND CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|aarch64le" AND CMAKE_CXX_COMPILER_ID MATCHES "GNU")
     if(CMAKE_BUILD_TYPE MATCHES Debug)
       message(Warning "Applying -Og optimization for aarch64 GCC debug build to workaround ICE")
     endif()
@@ -1196,7 +1204,7 @@ if(NOT MSVC)
   # aarch64 C++ stack unwinding uses frame-pointer chain walking, so frame
   # pointers must be present in all build types.  The cost is negligible on
   # aarch64 (31 GPRs vs x86-64's 16, so dedicating x29 rarely spills).
-  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|aarch64le")
     append_cxx_flag_if_supported("-fno-omit-frame-pointer" CMAKE_CXX_FLAGS)
   endif()
   append_cxx_flag_if_supported("-fno-math-errno" CMAKE_CXX_FLAGS)
@@ -1296,7 +1304,7 @@ if(ANDROID AND (NOT ANDROID_DEBUG_SYMBOLS))
   endif()
 endif()
 
-if(NOT APPLE AND UNIX)
+if(NOT APPLE AND NOT QNX AND UNIX)
   list(APPEND Caffe2_DEPENDENCY_LIBS dl)
 endif()
 
diff --git a/aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt b/aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt
index 19b7851b3e7..4e53f7e7fd3 100644
--- a/aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt
+++ b/aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt
@@ -54,13 +54,13 @@ if(NOT CMAKE_SYSTEM_PROCESSOR)
   else()
     message(FATAL_ERROR "CMAKE_SYSTEM_PROCESSOR is not defined")
   endif()
-elseif(NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^(i[3-6]86|x86_64|armv[5-8].*|aarch64|arm64)$")
+elseif(NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^(i[3-6]86|x86_64|armv[5-8].*|aarch64|arm64|aarch64le)$")
   message(FATAL_ERROR "Unrecognized CMAKE_SYSTEM_PROCESSOR = ${CMAKE_SYSTEM_PROCESSOR}")
 endif()
 
 if(NOT CMAKE_SYSTEM_NAME)
   message(FATAL_ERROR "CMAKE_SYSTEM_NAME not defined")
-elseif(NOT CMAKE_SYSTEM_NAME MATCHES "^(Darwin|Linux|Android)$")
+elseif(NOT CMAKE_SYSTEM_NAME MATCHES "^(Darwin|Linux|Android|QNX)$")
   message(FATAL_ERROR "Unrecognized CMAKE_SYSTEM_NAME = ${CMAKE_SYSTEM_NAME}")
 endif()
 
@@ -331,7 +331,7 @@ if(NOT TARGET clog)
     "${CONFU_DEPENDENCIES_BINARY_DIR}/clog")
   # We build static version of clog but a dynamic library may indirectly depend on it
   set_property(TARGET clog PROPERTY POSITION_INDEPENDENT_CODE ON)
-  if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID STREQUAL "QCC")
     target_compile_options(clog PRIVATE "-Wno-unused-result")
   endif()
 endif()
diff --git a/aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/math.h b/aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/math.h
index 82423546f9b..a66f93c512d 100644
--- a/aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/math.h
+++ b/aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/math.h
@@ -9,7 +9,7 @@
 #pragma once
 
 #include <stddef.h>
-#ifdef _MSC_VER
+#if  defined(_MSC_VER) || defined(__QNX__)
 #undef min
 #undef max
 #endif
diff --git a/c10/test/core/Scalar_test.cpp b/c10/test/core/Scalar_test.cpp
index d87cc051660..7416ba316cc 100644
--- a/c10/test/core/Scalar_test.cpp
+++ b/c10/test/core/Scalar_test.cpp
@@ -48,8 +48,10 @@ TEST(ScalarTest, Equality) {
   ASSERT_FALSE(Scalar(-1).equal(0xFFFFFFFFFFFFFFFF));
 }
 
+#ifndef __QNX__
 TEST(ScalarTest, LongsAndLongLongs) {
   Scalar longOne = 1L;
   Scalar longlongOne = 1LL;
   ASSERT_EQ(longOne.toInt(), longlongOne.toInt());
 }
+#endif
diff --git a/c10/test/util/exception_test.cpp b/c10/test/util/exception_test.cpp
index c260e96961a..934e6c2c4e9 100644
--- a/c10/test/util/exception_test.cpp
+++ b/c10/test/util/exception_test.cpp
@@ -34,7 +34,7 @@ TEST(ExceptionTest, TORCH_INTERNAL_ASSERT_DEBUG_ONLY) {
 }
 
 // On these platforms there's no assert
-#if !defined(__ANDROID__) && !defined(__APPLE__)
+#if !defined(__ANDROID__) && !defined(__APPLE__) && !defined(__QNX__)
 TEST(ExceptionTest, CUDA_KERNEL_ASSERT) {
   // This function always throws even in NDEBUG mode
   ASSERT_DEATH_IF_SUPPORTED({ CUDA_KERNEL_ASSERT(false); }, "Assert");
diff --git a/caffe2/CMakeLists.txt b/caffe2/CMakeLists.txt
index a6b6f0f7d1d..e5c73c2b705 100644
--- a/caffe2/CMakeLists.txt
+++ b/caffe2/CMakeLists.txt
@@ -561,7 +561,7 @@ if(CMAKE_COMPILER_IS_GNUCXX)
 endif()
 
 # Enable conditional FP16 arithmetic intrinsics
-if(CPU_AARCH64 AND LINUX)
+if(CPU_AARCH64 AND (LINUX OR QNX))
 set_source_files_properties(${TORCH_ROOT}/aten/src/ATen/native/BlasKernel.cpp PROPERTIES COMPILE_FLAGS "-march=armv8.2-a+fp16")
 endif()
 
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index 70c5d3de4cf..c22951b1b07 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -354,14 +354,14 @@ if(USE_NNPACK OR USE_PYTORCH_QNNPACK OR USE_XNNPACK)
       set(DISABLE_NNPACK_AND_FAMILY ON)
     endif()
   else()
-    if(NOT IOS AND NOT (CMAKE_SYSTEM_NAME MATCHES "^(Android|Linux|Darwin|Windows)$"))
+    if(NOT IOS AND NOT (CMAKE_SYSTEM_NAME MATCHES "^(Android|Linux|Darwin|Windows|QNX)$"))
       message(WARNING
         "Target platform \"${CMAKE_SYSTEM_NAME}\" is not supported in {Q/X}NNPACK. "
         "Supported platforms are Android, iOS, Linux, and macOS. "
         "Turn this warning off by USE_{Q/X}NNPACK=OFF.")
       set(DISABLE_NNPACK_AND_FAMILY ON)
     endif()
-    if(NOT IOS AND NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "^(i686|AMD64|x86_64|armv[0-9].*|arm64|aarch64)$"))
+    if(NOT IOS AND NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "^(i686|AMD64|x86_64|armv[0-9].*|arm64|aarch64|aarch64le)$"))
       message(WARNING
         "Target architecture \"${CMAKE_SYSTEM_PROCESSOR}\" is not supported in {Q/X}NNPACK. "
         "Supported architectures are x86, x86-64, ARM, and ARM64. "
diff --git a/scripts/build_mobile_qnx_arm.sh b/scripts/build_mobile_qnx_arm.sh
new file mode 100755
index 0000000000..9d86c3ab35
--- /dev/null
+++ b/scripts/build_mobile_qnx_arm.sh
@@ -0,0 +1,138 @@
+#!/bin/bash
+##############################################################################
+# Example command to build the mobile target.
+##############################################################################
+#
+# This script shows how one can build a libtorch library optimized for mobile
+# devices using host toolchain.
+
+set -e
+
+export BUILD_PYTORCH_MOBILE_WITH_HOST_TOOLCHAIN=1
+CAFFE2_ROOT="$( cd "$(dirname "$0")"/.. ; pwd -P)"
+
+if [ -z ${QNX_DIR+x} ]; then
+  QNX_DIR=$CAFFE2_ROOT
+fi
+
+if [ -z ${TEST+x} ]; then
+  TEST="OFF"
+fi
+
+echo "Bash: $(/bin/bash --version | head -1)"
+echo "Caffe2 path: $CAFFE2_ROOT"
+echo "Toolchain path: $QNX_DIR"
+
+CMAKE_ARGS=()
+CMAKE_ARGS+=("-DCMAKE_PREFIX_PATH=$(python -c 'import sysconfig; print(sysconfig.get_path("purelib"))')")
+CMAKE_ARGS+=("-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')")
+CMAKE_ARGS+=("-DBUILD_CUSTOM_PROTOBUF=OFF")
+#CMAKE_ARGS+=("-DBUILD_SHARED_LIBS=OFF")
+CMAKE_ARGS+=("-DBUILD_SHARED_LIBS=ON")
+
+# custom build with selected ops
+if [ -n "${SELECTED_OP_LIST}" ]; then
+  SELECTED_OP_LIST="$(cd $(dirname $SELECTED_OP_LIST); pwd -P)/$(basename $SELECTED_OP_LIST)"
+  echo "Choose SELECTED_OP_LIST file: $SELECTED_OP_LIST"
+  if [ ! -r ${SELECTED_OP_LIST} ]; then
+    echo "Error: SELECTED_OP_LIST file ${SELECTED_OP_LIST} not found."
+    exit 1
+  fi
+  CMAKE_ARGS+=("-DSELECTED_OP_LIST=${SELECTED_OP_LIST}")
+fi
+
+# If Ninja is installed, prefer it to Make
+if [ -x "$(command -v ninja)" ]; then
+  CMAKE_ARGS+=("-GNinja")
+fi
+
+# Don't build artifacts we don't need
+CMAKE_ARGS+=("-DBUILD_TEST=${TEST}")
+CMAKE_ARGS+=("-DINSTALL_TEST=OFF")
+CMAKE_ARGS+=("-DBUILD_BINARY=OFF")
+
+# If there exists env variable and it equals to 1, build lite interpreter.
+# Default behavior is to build full jit interpreter.
+# cmd:  BUILD_LITE_INTERPRETER=1 ./scripts/build_mobile.sh
+if [ "x${BUILD_LITE_INTERPRETER}" == "x1" ]; then
+  CMAKE_ARGS+=("-DBUILD_LITE_INTERPRETER=ON")
+else
+  CMAKE_ARGS+=("-DBUILD_LITE_INTERPRETER=OFF")
+fi
+if [ "x${TRACING_BASED}" == "x1" ]; then
+  CMAKE_ARGS+=("-DTRACING_BASED=ON")
+else
+  CMAKE_ARGS+=("-DTRACING_BASED=OFF")
+fi
+
+# Lightweight dispatch bypasses the PyTorch Dispatcher.
+if [ "${USE_LIGHTWEIGHT_DISPATCH}" == 1 ]; then
+  CMAKE_ARGS+=("-DUSE_LIGHTWEIGHT_DISPATCH=ON")
+  CMAKE_ARGS+=("-DSTATIC_DISPATCH_BACKEND=CPU")
+else
+  CMAKE_ARGS+=("-DUSE_LIGHTWEIGHT_DISPATCH=OFF")
+fi
+
+# Disable unused dependencies
+CMAKE_ARGS+=("-DUSE_ROCM=OFF")
+CMAKE_ARGS+=("-DUSE_CUDA=OFF")
+CMAKE_ARGS+=("-DUSE_ITT=OFF")
+CMAKE_ARGS+=("-DUSE_GFLAGS=OFF")
+CMAKE_ARGS+=("-DUSE_OPENCV=OFF")
+CMAKE_ARGS+=("-DUSE_LMDB=OFF")
+CMAKE_ARGS+=("-DUSE_LEVELDB=OFF")
+CMAKE_ARGS+=("-DUSE_MPI=OFF")
+CMAKE_ARGS+=("-DUSE_OPENMP=OFF")
+CMAKE_ARGS+=("-DUSE_MKLDNN=OFF")
+CMAKE_ARGS+=("-DUSE_NNPACK=OFF")
+CMAKE_ARGS+=("-DUSE_NUMPY=OFF")
+CMAKE_ARGS+=("-DUSE_BLAS=OFF")
+
+# Only toggle if VERBOSE=1
+if [ "${VERBOSE:-}" == '1' ]; then
+  CMAKE_ARGS+=("-DCMAKE_VERBOSE_MAKEFILE=1")
+fi
+
+# QNX config
+CMAKE_ARGS+=("-DXNNPACK_ENABLE_ASSEMBLY=OFF")
+
+CMAKE_ARGS+=("-DBUILD_QNX_ASM_FLAGS=-D_QNX_SOURCE -D__QNXNTO__")
+CMAKE_ARGS+=("-DBUILD_QNX_C_FLAGS=-D_QNX_SOURCE -D__QNXNTO__")
+CMAKE_ARGS+=("-DBUILD_QNX_CXX_FLAGS=-D_QNX_SOURCE -D__QNXNTO__")
+CMAKE_ARGS+=("-DBUILD_QNX_LINKER_FLAGS=-Wl,--build-id=md5")
+CMAKE_ARGS+=("-DCMAKE_TOOLCHAIN_FILE=$QNX_DIR/qnx.nto.toolchain.cmake")
+CMAKE_ARGS+=("-DCMAKE_SYSTEM_PROCESSOR=aarch64")
+CMAKE_ARGS+=("-DCMAKE_ASM_COMPILER_TARGET=gcc_ntoaarch64le")
+CMAKE_ARGS+=("-DCMAKE_C_COMPILER_TARGET=gcc_ntoaarch64le")
+CMAKE_ARGS+=("-DCMAKE_CXX_COMPILER_TARGET=gcc_ntoaarch64le")
+CMAKE_ARGS+=("-DCAFFE2_CUSTOM_PROTOC_EXECUTABLE=$QNX_DIR/host/protobuf/install/bin/protoc")
+CMAKE_ARGS+=("-DNATIVE_BUILD_DIR=$QNX_DIR/host/sleef")
+
+# User-specified CMake arguments go last to allow overridding defaults
+CMAKE_ARGS+=("$@")
+
+# Now, actually build the Android target.
+BUILD_ROOT=${BUILD_ROOT:-"$CAFFE2_ROOT/build_mobile"}
+INSTALL_PREFIX=${BUILD_ROOT}/install
+mkdir -p $BUILD_ROOT
+
+echo "${CMAKE_ARGS[@]}"
+
+cd $BUILD_ROOT
+cmake "$CAFFE2_ROOT" \
+    -DCMAKE_INSTALL_PREFIX=$INSTALL_PREFIX \
+    -DCMAKE_BUILD_TYPE=Release \
+    "${CMAKE_ARGS[@]}" \
+
+# Cross-platform parallel build
+if [ -z "$MAX_JOBS" ]; then
+  if [ "$(uname)" == 'Darwin' ]; then
+    MAX_JOBS=$(sysctl -n hw.ncpu)
+  else
+    MAX_JOBS=$(nproc)
+  fi
+fi
+
+echo "Will install headers and libs to $INSTALL_PREFIX for further project usage."
+cmake --build . --target install "-j${MAX_JOBS}" 2>> log.txt
+echo "Installation completed, now you can copy the headers/libs from $INSTALL_PREFIX to your project directory."
diff --git a/test/cpp/lite_interpreter_runtime/CMakeLists.txt b/test/cpp/lite_interpreter_runtime/CMakeLists.txt
index 6a2e6db6ea..2df57bfeb5 100644
--- a/test/cpp/lite_interpreter_runtime/CMakeLists.txt
+++ b/test/cpp/lite_interpreter_runtime/CMakeLists.txt
@@ -7,10 +7,16 @@ set(LITE_INTERPRETER_RUNTIME_TEST_DIR
   ${TORCH_ROOT}/test/cpp/lite_interpreter_runtime/test_mobile_profiler.cpp
 )
 
-add_library(backend_with_compiler_runtime SHARED
-        ${TORCH_ROOT}/test/cpp/jit/test_backend_compiler_lib.cpp
-        ${TORCH_ROOT}/torch/csrc/jit/backends/backend_interface.cpp
-        )
+if(QNX)
+  add_library(backend_with_compiler_runtime SHARED
+          ${TORCH_ROOT}/test/cpp/jit/test_backend_compiler_lib.cpp
+          )
+else()
+  add_library(backend_with_compiler_runtime SHARED
+          ${TORCH_ROOT}/test/cpp/jit/test_backend_compiler_lib.cpp
+          ${TORCH_ROOT}/torch/csrc/jit/backends/backend_interface.cpp
+          )
+endif()
 target_link_libraries(backend_with_compiler_runtime PRIVATE torch)
 
 add_executable(
diff --git a/test/cpp/lite_interpreter_runtime/test_lite_interpreter_runtime.cpp b/test/cpp/lite_interpreter_runtime/test_lite_interpreter_runtime.cpp
index f26cfcb2b5..b45ae3c2dd 100644
--- a/test/cpp/lite_interpreter_runtime/test_lite_interpreter_runtime.cpp
+++ b/test/cpp/lite_interpreter_runtime/test_lite_interpreter_runtime.cpp
@@ -16,9 +16,14 @@ namespace mobile {
 
 TEST(RunTimeTest, LoadAndForward) {
   // Load check in model: sequence.ptl
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "sequence.ptl";
+#else
   std::string filePath(__FILE__);
   auto testModelFile = filePath.substr(0, filePath.find_last_of("/\\") + 1);
   testModelFile.append("sequence.ptl");
+#endif
 
   //  sequence.ptl source code:
   //  class A(torch.nn.Module):
@@ -53,6 +58,10 @@ TEST(RunTimeTest, LoadAndForward) {
 }
 
 TEST(RunTimeTest, Delegate) {
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "delegate_test.ptl";
+#else
   std::string filePath(__FILE__);
   auto testModelFile = filePath.substr(0, filePath.find_last_of("/\\") + 1);
   // "delegate_test.ptl" is generated from test/cpp/jit/test_backend.cpp,
@@ -66,6 +75,7 @@ TEST(RunTimeTest, Delegate) {
   //        return x + h
   //  )");
   testModelFile.append("delegate_test.ptl");
+#endif
   auto mlm = _load_for_mobile(testModelFile);
   std::vector<IValue> inputs;
   inputs.emplace_back(2.0 * at::ones({}));
@@ -76,6 +86,10 @@ TEST(RunTimeTest, Delegate) {
 }
 
 TEST(RunTimeTest, DelegateException) {
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "delegated_submodule_with_debug_info.ptl";
+#else
   std::string filePath(__FILE__);
   auto testModelFile = filePath.substr(0, filePath.find_last_of("/\\") + 1);
   /*
@@ -138,6 +152,7 @@ TEST(RunTimeTest, DelegateException) {
    *
    */
   testModelFile.append("delegated_submodule_with_debug_info.ptl");
+#endif
   auto mlm = _load_for_mobile(testModelFile);
   std::vector<IValue> inputs;
   inputs.emplace_back(torch::rand({2, 4}));
diff --git a/test/cpp/lite_interpreter_runtime/test_mobile_profiler.cpp b/test/cpp/lite_interpreter_runtime/test_mobile_profiler.cpp
index 2ebb247bd2..189f526791 100644
--- a/test/cpp/lite_interpreter_runtime/test_mobile_profiler.cpp
+++ b/test/cpp/lite_interpreter_runtime/test_mobile_profiler.cpp
@@ -45,15 +45,24 @@ bool checkMetaData(
 } // namespace
 
 TEST(MobileProfiler, ModuleHierarchy) {
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "to_be_profiled_module.ptl";
+#else
   auto testModelFile = torch::testing::getResourcePath(
       "test/cpp/lite_interpreter_runtime/to_be_profiled_module.ptl");
+#endif
 
   std::vector<IValue> inputs;
   inputs.emplace_back(at::rand({64, 64}));
   inputs.emplace_back(at::rand({64, 64}));
   std::string trace_file_name("/tmp/test_trace.trace");
 
+#if defined(__QNX__)
+  mobile::Module bc = _load_for_mobile(testModelFile);
+#else
   mobile::Module bc = _load_for_mobile(testModelFile.string());
+#endif
   {
     KinetoEdgeCPUProfiler profiler(
         bc,
@@ -99,15 +108,24 @@ TEST(MobileProfiler, ModuleHierarchy) {
 }
 
 TEST(MobileProfiler, Backend) {
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "test_backend_for_profiling.ptl";
+#else
   auto testModelFile = torch::testing::getResourcePath(
       "test/cpp/lite_interpreter_runtime/test_backend_for_profiling.ptl");
+#endif
 
   std::vector<IValue> inputs;
   inputs.emplace_back(at::rand({64, 64}));
   inputs.emplace_back(at::rand({64, 64}));
   std::string trace_file_name("/tmp/test_trace_backend.trace");
 
+#if defined(__QNX__)
+  mobile::Module bc = _load_for_mobile(testModelFile);
+#else
   mobile::Module bc = _load_for_mobile(testModelFile.string());
+#endif
   {
     KinetoEdgeCPUProfiler profiler(
         bc,
@@ -133,15 +151,24 @@ TEST(MobileProfiler, Backend) {
 }
 
 TEST(MobileProfiler, BackendMemoryEvents) {
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "test_backend_for_profiling.ptl";
+#else
   auto testModelFile = torch::testing::getResourcePath(
       "test/cpp/lite_interpreter_runtime/test_backend_for_profiling.ptl");
+#endif
 
   std::vector<IValue> inputs;
   inputs.emplace_back(at::rand({64, 64}));
   inputs.emplace_back(at::rand({64, 64}));
   std::string trace_file_name("/tmp/test_trace_backend_memory.trace");
 
+#if defined(__QNX__)
+  mobile::Module bc = _load_for_mobile(testModelFile);
+#else
   mobile::Module bc = _load_for_mobile(testModelFile.string());
+#endif
   {
     mobile::KinetoEdgeCPUProfiler profiler(
         bc,
@@ -165,8 +192,13 @@ TEST(MobileProfiler, BackendMemoryEvents) {
 }
 
 TEST(MobileProfiler, ProfilerEvent) {
+#if defined(__QNX__)
+  // Don't compile an absolute path when building for a target.
+  auto testModelFile = "test_backend_for_profiling.ptl";
+#else
   auto testModelFile = torch::testing::getResourcePath(
       "test/cpp/lite_interpreter_runtime/test_backend_for_profiling.ptl");
+#endif
 
   std::vector<IValue> inputs;
   inputs.emplace_back(at::rand({64, 64}));
@@ -177,7 +209,11 @@ TEST(MobileProfiler, ProfilerEvent) {
       torch::profiler::ProfilerPerfEvents.begin(),
       torch::profiler::ProfilerPerfEvents.end());
 
+#if defined(__QNX__)
+  mobile::Module bc = _load_for_mobile(testModelFile);
+#else
   mobile::Module bc = _load_for_mobile(testModelFile.string());
+#endif
   {
     // Bail if something goes wrong here
     try {
diff --git a/torch/csrc/DataLoader.cpp b/torch/csrc/DataLoader.cpp
index a0c668043b0..0a95b96a4d2 100644
--- a/torch/csrc/DataLoader.cpp
+++ b/torch/csrc/DataLoader.cpp
@@ -56,7 +56,11 @@ static inline void setSignalHandler(
     struct sigaction* old_sa_ptr) {
   struct sigaction sa {};
   sa.sa_sigaction = handler;
+#ifndef SA_RESTART
+  sa.sa_flags = SA_SIGINFO | SA_NOCLDSTOP | SA_NODEFER;
+#else
   sa.sa_flags = SA_RESTART | SA_SIGINFO | SA_NOCLDSTOP | SA_NODEFER;
+#endif
   if (sigemptyset(&sa.sa_mask) != 0 ||
       sigaction(signal, &sa, old_sa_ptr) != 0) {
     std::ostringstream oss;
diff --git a/test/mobile/lightweight_dispatch/CMakeLists.txt b/test/mobile/lightweight_dispatch/CMakeLists.txt
index 5ab3232f6a..b83fc7528f 100644
--- a/test/mobile/lightweight_dispatch/CMakeLists.txt
+++ b/test/mobile/lightweight_dispatch/CMakeLists.txt
@@ -8,7 +8,11 @@ add_executable(test_codegen_unboxing
   ${TEST_ROOT}/test_codegen_unboxing.cpp
 )
 
-target_include_directories(test_codegen_unboxing PRIVATE ${ATen_CPU_INCLUDE})
+if(QNX)
+  target_include_directories(test_codegen_unboxing PRIVATE ${ATen_CPU_INCLUDE} ${TORCH_ROOT}/torch/csrc/api/include)
+else()
+  target_include_directories(test_codegen_unboxing PRIVATE ${ATen_CPU_INCLUDE})
+endif()
 
 target_compile_definitions(test_codegen_unboxing PRIVATE USE_GTEST)
 
diff --git a/torch/csrc/utils/byte_order.h b/torch/csrc/utils/byte_order.h
index d586270fbfc..e852338059f 100644
--- a/torch/csrc/utils/byte_order.h
+++ b/torch/csrc/utils/byte_order.h
@@ -16,6 +16,11 @@
 #define thp_bswap16(x) bswap16(x)
 #define thp_bswap32(x) bswap32(x)
 #define thp_bswap64(x) bswap64(x)
+#elif defined (__QNX__)
+#include <gulliver.h>
+#define thp_bswap16(x) ENDIAN_SWAP16(x)
+#define thp_bswap32(x) ENDIAN_SWAP32(x)
+#define thp_bswap64(x) ENDIAN_SWAP64(x)
 #elif defined(__APPLE__)
 #include <libkern/OSByteOrder.h>
 #define thp_bswap16(x) OSSwapInt16(x)
diff --git a/torch/headeronly/macros/Macros.h b/torch/headeronly/macros/Macros.h
index cef99df3f56..bc71c4b4cd9 100644
--- a/torch/headeronly/macros/Macros.h
+++ b/torch/headeronly/macros/Macros.h
@@ -545,6 +545,40 @@ __host__ __device__
     abort();                     \
   }
 #else
+#if defined(__QNX__)
+#define CUDA_KERNEL_ASSERT(cond)                                         \
+  if (C10_UNLIKELY(!(cond))) {                                           \
+    __assert(                                                       \
+        #cond, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
+  }
+#define CUDA_KERNEL_ASSERT_MSG(cond, msg)                              \
+  if (C10_UNLIKELY(!(cond))) {                                         \
+    __assert(                                                     \
+        msg, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
+  }
+#define CUDA_KERNEL_ASSERT_PRINTF(cond, msg, ...)                        \
+  if (C10_UNLIKELY(!(cond))) {                                           \
+    printf(                                                            \
+        "[CUDA_KERNEL_ASSERT] " __FILE__ ":" C10_STRINGIZE(            \
+            __LINE__) ": %s: block: [%d,%d,%d], thread: [%d,%d,%d]: "  \
+            "Assertion failed: `" #cond "`: " msg "\n",                \
+        __func__,                                                      \
+        blockIdx.x,                                                    \
+        blockIdx.y,                                                    \
+        blockIdx.z,                                                    \
+        threadIdx.x,                                                   \
+        threadIdx.y,                                                   \
+        threadIdx.z,                                                   \
+        ##__VA_ARGS__); \
+    __assert(                                                       \
+        #cond, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
+  }
+#define SYCL_KERNEL_ASSERT(cond)                                         \
+  if (C10_UNLIKELY(!(cond))) {                                           \
+    __assert(                                                       \
+        #cond, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
+  }
+#else
 #define CUDA_KERNEL_ASSERT(cond)                                         \
   if (C10_UNLIKELY(!(cond))) {                                           \
     __assert_fail(                                                       \
@@ -577,6 +611,7 @@ __host__ __device__
     __assert_fail(                                                       \
         #cond, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
   }
+#endif // __QNX__
 #endif //  C10_USE_ROCM_KERNEL_ASSERT && USE_ROCM
 #endif // __APPLE__
 
diff --git a/torch/lib/libshm/CMakeLists.txt b/torch/lib/libshm/CMakeLists.txt
index 688e7e79291..80b6bf35713 100644
--- a/torch/lib/libshm/CMakeLists.txt
+++ b/torch/lib/libshm/CMakeLists.txt
@@ -66,6 +66,9 @@ else()
   # we need to link directly to c10 here otherwise we miss symbols
   target_link_libraries(torch_shm_manager PRIVATE shm c10)
 endif()
+if(CMAKE_SYSTEM_NAME STREQUAL "QNX")
+  target_link_libraries(torch_shm_manager PRIVATE socket)
+endif()
 set_target_properties(torch_shm_manager PROPERTIES
   INSTALL_RPATH "${_rpath_portable_origin}/../lib")
 
